token-lending/program/src/state/reserve.rs

```diff
@@ -74,265 +108,200 @@ impl Reserve {
-            };
-
-            Ok(LiquidateResult {
-                withdraw_amount,
-                repay_amount,
-                settle_amount,
-            })
-        } else {
-            Err(LendingError::LiquidationTooSmall.into())
+    /// Update borrow rate and accrue interest
+    pub fn accrue_interest(&mut self, current_slot: Slot) -> ProgramResult {
+        let slots_elapsed = self.last_update.slots_elapsed(current_slot)?;
+        if slots_elapsed > 0 {
+            let current_borrow_rate = self.current_borrow_rate()?;
+            self.liquidity
+                .compound_interest(current_borrow_rate, slots_elapsed)?;
+            self.last_update.update_slot(current_slot);
         }
+        Ok(())
     }
 
-    /// Create new loan
-    pub fn create_loan(
+    /// Borrow liquidity up to a maximum market value
+    pub fn borrow_liquidity(
         &self,
-        token_amount: u64,
-        token_amount_type: BorrowAmountType,
-        token_converter: impl TokenConverter,
-        borrow_amount_token_mint: &Pubkey,
-    ) -> Result<LoanResult, ProgramError> {
-        let (borrow_amount, mut collateral_amount) = match token_amount_type {
-            BorrowAmountType::CollateralDepositAmount => {
-                let collateral_amount = token_amount;
-                let borrow_amount =
-                    self.allowed_borrow_for_collateral(collateral_amount, token_converter)?;
-                if borrow_amount == 0 {
-                    return Err(LendingError::InvalidAmount.into());
+        liquidity_amount: u64,
+        liquidity_amount_type: AmountType,
+        max_borrow_value: Decimal,
+    ) -> Result<BorrowLiquidityResult, ProgramError> {
+        match liquidity_amount_type {
+            AmountType::ExactAmount => {
+                let receive_amount = liquidity_amount;
+                let borrow_amount = Decimal::from(receive_amount);
+                let (borrow_fee, host_fee) = self
+                    .config
+                    .fees
+                    .calculate_borrow_fees(borrow_amount, FeeCalculation::Exclusive)?;
+
+                let borrow_amount = borrow_amount.try_add(borrow_fee.into())?;
+                let borrow_value = borrow_amount.try_mul(self.liquidity.median_price)?;
+                if borrow_value > max_borrow_value {
+                    return Err(LendingError::BorrowTooLarge.into());
                 }
-                (borrow_amount, collateral_amount)
+
+                Ok(BorrowLiquidityResult {
+                    borrow_amount,
+                    receive_amount,
+                    borrow_fee,
+                    host_fee,
+                })
             }
-            BorrowAmountType::LiquidityBorrowAmount => {
-                let borrow_amount = token_amount;
-                let collateral_amount = self.required_collateral_for_borrow(
+            AmountType::PercentAmount => {
+                let borrow_pct = Rate::from_percent(liquidity_amount as u8);
```

Comment:
So, liquidity amount is checked by the caller to be in the range `(0,100]` initially. I think the percent amount is important because clients cannot precisely determine total borrowing power at the time of the borrow. The most they can do is estimate it, and then it can change in flight. If a user tries to borrow what a client says is 100%, but then the value of the deposits/borrows is refreshed, the amount sent to the program may be insufficient to cover the borrow + fees, and it will error because a precise borrow is supposed to fail hard if the user can't receive the tokens they ask for.

Reply:
Ah, didn't see the range limit, thanks.

> The most they can do is estimate it, and then it can change in flight.

Yes, this is true. But in Solana, the latency is low so in-flight changes are bound to be minimal. Your example of borrowing 100% is not very realistic because I don't think people will actually want to do that. I think minor slippage is ok because usually people expect to borrow a specific amount of an asset and use % as a rough guideline. Also, I haven't seen other protocols support percentage inputs.

Reply:
I think this API was @bartosz-lipinski's idea, so I might have to get him to defend it haha

But thinking this through some more, clients frequently show percent sliders, even if the protocol can't handle the input. It would be nice for end users if it could.

What about when withdrawing collateral or repaying? Let's say you go to repay an entire loan, and you provide a fixed amount. If the value of the loan decreases on refresh, you won't overpay, but if increases, you'll pay slightly too little and have to repeat it. Clients can handle this, but this just moves the complexity to them.

I imagine liquidations are handled by programs more often than not, but it's kind of a similar thing -- a liquidator is probably interest in liquidating as much as they can. This gives them the ability to do that, but they could also set slippage or a fixed maximum to pay.

Reply:
> Let's say you go to repay an entire loan, and you provide a fixed amount. If the value of the loan decreases on refresh, you won't overpay, but if increases, you'll pay slightly too little and have to repeat it.

Yes, this is really valid edge case, this is why some protocols provide a special value you can use to pay off the full loan (for example, setting amount to the max u64)

> But thinking this through some more, clients frequently show percent sliders, even if the protocol can't handle the input. It would be nice for end users if it could.

Yup, agreed. It's nice for end users so clients can choose to provide that abstraction

> a liquidator is probably interest in liquidating as much as they can.

Good point. Again, I think this can be handled with a special value to indicate the max amount should be liquidated. All other cases should be fine with a fixed amount, I think.

Reply:
Done.